DSCResources/MSFT_xDisk/MSFT_xDisk.psm1

```diff
@@ -28,96 +40,167 @@ function Set-TargetResource
+        Throw "Disk Set-TargetResource failed with the following error: '$($Error[0])'"
     }
-    
-    $global:DSCMachineStatus = 1    
 }
 
 function Test-TargetResource
 {
     [OutputType([System.Boolean])]
+    [cmdletbinding()]
     param
     (
         [parameter(Mandatory)]
         [uint32] $DiskNumber,
 
-        [string] $DriveLetter
+        [parameter(Mandatory)]
+        [string] $DriveLetter,
+
+        [UInt64] $Size,
+        [string] $FSLabel
     )
 
     Write-Verbose -Message "Checking if disk number '$($DiskNumber)' is initialized..."
-    $disk = Get-Disk -Number $DiskNumber
-    if (-not $disk)
+    $Disk = Get-Disk -Number $DiskNumber -ErrorAction SilentlyContinue
+
+    if (-not $Disk)
     {
-        throw "Disk number '$($DiskNumber)' does not exist."
+        Write-Verbose "Disk number '$($DiskNumber)' was not found."
+        return $false
     }
-    if ($disk.PartitionStyle -ne "RAW")
+
+    if ($Disk.IsOffline -eq $true)
     {
-        Write-Verbose "Disk number '$($DiskNumber)' has already been initialized."
+        Write-Verbose 'Disk is not Online'
+        return $false
+    }
+    
+    if ($Disk.IsReadOnly -eq $true)
+    {
+        Write-Verbose 'Disk set as ReadOnly'
+        return $false
+    }
 
-        $driveLetterFromDisk = $disk | Get-Partition | Where-Object { $_.DriveLetter -ne "`0" } | Select-Object -ExpandProperty DriveLetter
-        if ($DriveLetter -ne "" -and $DriveLetter -ne $driveLetterFromDisk)
-        {
-            write-verbose "Disk number '$($DiskNumber)' has an unexpected drive letter. Expected: $DriveLetter. Actual: $driveLetterFromDisk."
-            return $false
-        }
-        if ($disk.IsOffline -eq $true) 
+    if ($Disk.PartitionStyle -ne "GPT")
+    {
+        Write-Verbose "Disk '$($DiskNumber)' is initialised with '$($Disk.PartitionStyle)' partition style"
+        return $false
+    }
+
+    $Partition = Get-Partition -DriveLetter $DriveLetter -ErrorAction SilentlyContinue
+    if (-not $Partition.DriveLetter -eq $DriveLetter)
+    {
+        Write-Verbose "Drive $DriveLetter was not found"
+        return $false
+    }
+
+    # Drive size
+    if ($Size)
+    {
+        if ($Partition.Size -ne $Size)
         {
-            write-verbose "Disk is set Offline."
+            Write-Verbose "Drive $DriveLetter size does not match defined value"
             return $false
         }
-        if ($disk.IsReadOnly -eq $true) 
+    }
+
+    # Volume label
+    if (-not [string]::IsNullOrEmpty($FSLabel))
+    {
+        $Label = Get-Volume -DriveLetter $DriveLetter -ErrorAction SilentlyContinue | Select-Object -ExpandProperty FileSystemLabel
+        if ($Label -ne $FSLabel)
         {
-            write-verbose "Disk is set ReadOnly."
+            Write-Verbose "Volume $DriveLetter label does not match defined value"
```

Comment:
same as above..
It's useful to display what's expected and current value of label